File: 4. longest_common_prefix.py
# The Longest Common Prefix problem requires finding the longest prefix string that is common among all the strings in a given array of lowercase English words. A prefix is defined as the starting portion of a string. You must compare the strings and determine the longest sequence of characters from the beginning that appears in every string in the array. If no common prefix exists among all the strings, the function should return an empty string (""). The solution must handle cases where the array contains one or more strings, and each string may have different lengths.
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs:
            return ""

        prefix = strs[0]

        for string in strs[1:]:
            while not string.startswith(prefix):
                prefix = prefix[:-1]
                if prefix == "":
                    return ""

        return prefix

File: 4. longest_common_prefix.js
// Given an array of strings, your task is to find the longest common prefix shared among all the strings and return it. A prefix is the beginning part of a string, so for example if you have the array ["flower", "flow", "flight"], the longest common prefix is "fl" because all three strings start with "fl" but diverge after that. If there is no common prefix at all among the strings, like in the case of ["dog", "racecar", "car"] where no single starting character is shared across all words, you should return an empty string "". The key challenge here is to find the longest prefix that every single string in the array starts with — not just some of them, but all of them. Keep in mind that the array can contain up to 200 strings and each string can be up to 200 characters long, and the strings will only consist of lowercase English letters. Edge cases to watch out for include an empty string being present in the array, which would immediately make the common prefix empty, as well as an array containing only one string, in which case the entire string itself is the answer.
var longestCommonPrefix = function(strs){
    if(strs.length === 0) 
        return "";

    let prefix = strs[0];

    for(let i = 1; i <strs.length; i++){
        while(strs[i].indexOf(prefix) !== 0){
            prefix = prefix.slice(0, -1);
            if(prefix === "")
                return "";
        }
    }

    return prefix;
};
